CORS SPI

dteleguin opened the pull request:
Closes #25446

mposolda commented:
Thanks for this draft!

My vote is to have `CorsProvider` and related classes in the `keycloak-server-spi-private` rather than adding them to `keycloak-server-spi`. The `keycloak-server-spi` has restrictions on backwards compatibility and it is very hard to change those classes due to this and IMO there is no need to have `Cors` class exposed here.

This means that instead of `session.cors(request)`, you would have something like `session.getProvider(CorsProvider.class).getCors(request)`, which IMO is completely sufficient? Also it would allow that there is no need to move `RestEasy` and `RestEasyProvider` classes?

pedroigor commented on common/src/main/java/org/keycloak/common/util/ResteasyProvider.java:
```diff
@@ -0,0 +1,32 @@
+
+package org.keycloak.common.util;
+
+public interface ResteasyProvider {
```
I hope we can get rid of this interface one day :)

pedroigor reviewed (commented):
@dteleguin The changes look good to me if others agree the SPI path is OK. But why do we want to allow customizations to how we handle CORS? 

I would say that most of the time we want to enhance our CORS support instead of people making customizations. If you really need to change configuration and how it works it should be done by configuration and the administration console rather than passing options when starting the server.

I think it provides a better UX while at the same time avoids introducing yet another SPI.

mposolda commented:
I see the use-case of customization of Cors SPI as not that important. However I do see the fact that `CorsProvider` can be in `keycloak-server-spi-private` when the implementation can be still kept in `services` as useful as Cors can then be better used from other modules without need to have dependency on `keycloak-services`. So for me, the first point from the Motivation section https://github.com/keycloak/keycloak/issues/25446  sounds like most important one.

pedroigor commented:
@mposolda Yeah, if we go with the SPI approach it is fine for me. But if we can avoid another SPI (loading/initing it during build/boot) but just a common API for handling CORS, it would be better IMO.

The problem I'm pointing out is very similar to the `security-headers` SPI. Do we really need that SPI or we could just live with customizing it through the admin console as we do today?

Otherwise, the changes LGTM as the changes are very self-contained.

pedroigor commented:
@mposolda @dteleguin How to proceed with this one?

dteleguin commented:
> @dteleguin The changes look good to me if others agree the SPI path is OK. But why do we want to allow customizations to how we handle CORS?
> 
> I would say that most of the time we want to enhance our CORS support instead of people making customizations.

To me, allowing customizations is equally important. Let me give an example; in a big k8s appliance including Keycloak, we might need to easily add/remove an application that would require a valid CORS origin. (That could be a test/demo application or similar.) The origin would be defined in a k8s environment, and we would want to minimize our efforts and avoid cycling through Keycloak realms and clients adding/removing web origins. This is a real world example which is pretty legitimate, but probably too specific to be upstreamed in Keycloak. Currently, we're solving that with a JAX-RS filter, but the solution is pretty verbose, and could be way nicer and simpler with CORS SPI.

> If you really need to change configuration and how it works it should be done by configuration and the administration console rather than passing options when starting the server.
> 
> I think it provides a better UX while at the same time avoids introducing yet another SPI.

Again, I would say supporting configuration via environment variables is equally important. Another example is #12682 which is all about supporting different tracing systems. IMO, tracing is the concept that definitely belongs to the DevOps/cloud world; and in that world, using env vars has become a de facto standard (whether we like it or not :man_shrugging:).

As for configuring CORS-related parameters via admin console, I don't think there are any conflicts here. Ideally, we could support realm-level params via admin console and global params via env vars. On the other hand, I remember @stianst 's [vision](https://github.com/keycloak/keycloak/issues/12682#issuecomment-1440352215) is to get rid of Realm settings -> Security defenses, so the global config could become the only option one day.

pedroigor commented:
@dteleguin Fair enough. Are you planning to mark it read for review and merged before KC 24?

dteleguin commented:
> @dteleguin Fair enough. Are you planning to mark it read for review and merged before KC 24?

Yes, that's the plan. Hope it's not too late for KC 24, need a couple of days to sort the outstanding issues.

dteleguin committed:
CORS SPI

Closes #25446

Signed-off-by: Dmitry Telegin <demetrio@carretti.pro>

dteleguin commented:
I have slightly reworked the original PR. The highlights are:
* moved SPI classes to `server-spi-private`, as suggested by @mposolda;
* Resteasy and ResteasyProvider remain in `common`, as they still need to be visible from `server-spi-private` in order to keep the old API working;
* minor improvements:
  * some of Cors `build()` methods now return a boolean indicating whether CORS has been applied, in order to improve logging and diagnostics;
  * added checks for `HttpRequest` and `ResponseBuilder` presence.

The CORS APIs now look like this:
```java
// old API
Cors cors = Cors.add(request);
Cors cors = Cors.add(request, response);

// new API
Cors cors = session.getProvider(Cors.class); // request object will be obtained from KeycloakContext, should cover the absolute majority of cases
Cors cors = session.getProvider(Cors.class).request(request); // override request object
Cors cors = session.getProvider(Cors.class).builder(builder); // set builder
Cors cors = session.getProvider(Cors.class).request(request).builder(builder); // override both, should be rare
```

The PR is ready for review now. @pedroigor

dteleguin commented:
Some final thoughts on why it might be beneficial to have an SPI for CORS:
* This is essentially how we pass implementation classes between modules and ensure loose coupling. If we resort to just interface + impl, providers like Realm REST Resource will need to depend on `keycloak-services` and to know the implementing class, in order to use Cors;
* Some real-world use cases require alternate CORS implementation that we wouldn't want to upstream; see the above example (overriding/augmenting CORS-related configuration using custom, deployment-specific sources);
* the SPI also provides a config namespace (`kc.spi-cors-default`) which we can further use to define global CORS settings. These could even be mapped to CLI shortcuts like `--cors-...`.

thomasdarimont commented on server-spi-private/src/main/java/org/keycloak/services/cors/Cors.java:
```diff
@@ -0,0 +1,92 @@
+    public static final String DEFAULT_ALLOW_HEADERS = "Origin, Accept, X-Requested-With, Content-Type, Access-Control-Request-Method, Access-Control-Request-Headers, DPoP";
+
+    public static final String ORIGIN_HEADER = "Origin";
+    public static final String AUTHORIZATION_HEADER = "Authorization";
```
How about using `jakarta.ws.rs.core.HttpHeaders#AUTHORIZATION` instead?

dteleguin replied:
Good point. In fact, all that hardcoded header stuff looks a bit messy - and I'm planning to address that in #12682 as it's directly related.

thomasdarimont commented on server-spi-private/src/main/java/org/keycloak/services/cors/Cors.java:
```diff
@@ -0,0 +1,92 @@
+    public static final String DEFAULT_ALLOW_METHODS = "GET, HEAD, OPTIONS";
+    public static final String DEFAULT_ALLOW_HEADERS = "Origin, Accept, X-Requested-With, Content-Type, Access-Control-Request-Method, Access-Control-Request-Headers, DPoP";
+
+    public static final String ORIGIN_HEADER = "Origin";
```
I see that this interface declares some constants for known CORS headers.

Idea: There is also org.keycloak.adapters.CorsHeaders in adapter-core which does the same, perhaps we could create a CorsHeaders interface in keycloak-core and use that in the adapter-core the Cors implementation.

dteleguin replied:
(same as above)

thomasdarimont commented on server-spi-private/src/main/java/org/keycloak/services/cors/CorsSpi.java:
```diff
@@ -0,0 +1,53 @@
+
+    @Override
+    public String getName() {
+        return SPI_ID;
```
Nitpick: just inline the constant in getName() as the constant is private anyway.

dteleguin replied:
Also good point, I've checked KC SPI classes and they introduce a (public) static constant only if it's needed elsewhere (e.g. tests). We can add it later if needed.

thomasdarimont commented on services/src/main/java/org/keycloak/services/cors/DefaultCorsImpl.java:
```diff
@@ -36,26 +38,9 @@
-
-    public static final String ORIGIN_HEADER = "Origin";
-    public static final String AUTHORIZATION_HEADER = "Authorization";
+public class DefaultCorsImpl implements Cors {
```
Nitpick: I'd drop the impl and go with `DefaultCors`

dteleguin replied:
I was thinking about that too. "Impl" seemed more natural as CORS is a technology and not just object. But KC convention is different (`DefaultKeycloakSession` etc.), so let's just follow it :)

thomasdarimont reviewed (changes requested):
Great SPI that will make it easier to configure cors!

I have some small suggestions, mostly nitpicks - feel free to adapt or ignore :)

pedroigor commented:
@mposolda LGTM.

dteleguin committed:
CORS SPI - code review

Signed-off-by: Dmitry Telegin <demetrio@carretti.pro>

pedroigor commented:
@mposolda @thomasdarimont Merging as it is needed by https://github.com/keycloak/keycloak/pull/26251.

We can have follow-ups (@dteleguin already have some) to improve more of it.

dteleguin commented:
thanks everyone! :partying_face:

dteleguin commented:
I have tried to address the HTTP header constants issue as part of #27334. Could you please take a look @thomasdarimont @pedroigor ?
